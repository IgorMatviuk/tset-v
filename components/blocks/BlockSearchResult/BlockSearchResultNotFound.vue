<template>
  <div class="no-search-result">
    <BContainer>
      <SearchBox
        :with-label="true"
        :initial-value="query"
        class="no-search-result__box"
      />
      <div class="no-search-result__wrapper">
        <svg-icon
          name="dress"
          class="no-search-result__icon"
          width="107"
          height="107"
        />
        <div class="no-search-result__title">
          {{ $t('searchPage.notFoundTitle') }}
        </div>
        <div class="no-search-result__description">
          {{ $t('searchPage.notFoundDescription') }}
        </div>
        <BButton
          variant="primary"
          :to="backToShoppingLink"
          class="no-search-result__to-catalog"
        >
          {{ $t('buttons.catalog') }}
        </BButton>
      </div>
    </BContainer>
  </div>
</template>

<script>
import SearchBox from '@/components/functional/SearchBox'

export default {
  components: {
    SearchBox
  },
  props: {
    query: {
      type: String,
      required: true
    }
  },
  computed: {
    backToShoppingLink() {
      return this.$i18n.path(this.$store.getters['catalog/defaultCategorySlug'])
    }
  }
}
</script>

<style lang="scss">
.no-search-result {
  min-height: calc(100vh - 430px);
  padding-top: 8px;
  padding-bottom: 50px;

  &__box {
    max-width: 690px;
    margin-right: auto;
    margin-left: auto;

    .search-box {
      &__label {
        display: block;
        margin-bottom: 20px;
        font-size: 16px;
        text-align: center;
      }
    }
  }

  &__wrapper {
    padding-top: 57px;
    text-align: center;
  }

  &__icon {
    width: 107px;
    height: 107px;
    margin-bottom: 24px;
    color: $primary-light-5;
  }

  &__title {
    max-width: 535px;
    margin-right: auto;
    margin-bottom: 18px;
    margin-left: auto;
    font-size: 16px;
    font-weight: 700;
    line-height: 20px;
    letter-spacing: 0.02em;
  }

  &__description {
    max-width: 535px;
    margin-right: auto;
    margin-bottom: 31px;
    margin-left: auto;
    font-size: 13px;
    line-height: 16px;
    color: $gray-dark-2;
  }

  @include media-breakpoint-down(xl) {
    &__box {
      max-width: 570px;
    }

    &__wrapper {
      padding-top: 50px;
    }

    &__icon {
      width: 90px;
      height: 90px;
      margin-bottom: 22px;
    }

    &__description {
      margin-bottom: 25px;
    }
  }

  @include media-breakpoint-down(lg) {
    min-height: calc(100vh - 480px);

    &__box {
      max-width: 450px;

      .search-box {
        &__label {
          margin-bottom: 15px;
        }
      }
    }

    &__wrapper {
      padding-top: 40px;
    }

    &__icon {
      width: 80px;
      height: 80px;
      margin-bottom: 20px;
    }

    &__title {
      max-width: 450px;
      margin-bottom: 10px;
      font-size: 14px;
      line-height: 18px;
    }

    &__description {
      max-width: 450px;
      margin-bottom: 20px;
      font-size: 12px;
      line-height: 15px;
    }
  }

  @include media-breakpoint-down(md) {
    min-height: calc(100vh - 590px);

    &__box {
      .search-box {
        &__lable {
          margin-bottom: 18px;
        }
      }
    }

    &__wrapper {
      padding-top: 38px;
    }

    &__icon {
      display: none;
    }

    &__title {
      margin-bottom: 10px;
    }

    &__descriptions {
      margin-bottom: 20px;
    }
  }

  @include media-breakpoint-down(xs) {
    min-height: calc(100vh - 510px);
    padding-top: 25px;
    padding-bottom: 70px;

    &__box {
      max-width: 310px;

      .search-box {
        &__label {
          margin-bottom: 19px;
        }
      }
    }

    &__wrapper {
      padding-top: 40px;
    }

    &__title {
      margin-bottom: 15px;
    }

    &__description {
      margin-bottom: 25px;
    }
  }
}
</style>
