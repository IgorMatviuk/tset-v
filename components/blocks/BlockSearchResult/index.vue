<template>
  <div class="search-result">
    <BContainer>
      <div class="search-result__header">
        <SearchBox
          :initial-value="query"
          :with-label="true"
          class="search-result__box"
        />
        <PagePagination
          v-if="totalPages > 1"
          class="search-result__top-pagination"
          :current-page="currentPage"
          :total-rows="totalPages"
          @page-change="onPageChange"
        />
      </div>
      <div class="search-result__row">
        <div
          v-for="product in products"
          :key="product.id"
          class="search-result__item"
        >
          <ProductPreview :item="product" />
        </div>
        <PagePagination
          v-if="totalPages > 1"
          class="search-result__pagination"
          :current-page="currentPage"
          :total-rows="totalPages"
          @page-change="onPageChange"
        />
      </div>
    </BContainer>
  </div>
</template>

<script>
import { createNamespacedHelpers } from 'vuex'

import SearchBox from '@/components/functional/SearchBox'
import ProductPreview from '@/components/shared/ProductPreview'
import PagePagination from '@/components/shared/PagePagination'

import searchPageChange from '@/mixins/search-page-change'

const { mapState } = createNamespacedHelpers('search')

export default {
  components: {
    SearchBox,
    ProductPreview,
    PagePagination
  },
  mixins: [searchPageChange],
  props: {
    products: {
      type: Array,
      required: true
    },
    query: {
      type: String,
      required: true
    }
  },
  computed: {
    ...mapState(['currentPage', 'totalPages'])
  }
}
</script>

<style lang="scss">
.search-result {
  padding-top: 8px;
  padding-bottom: 75px;
  overflow: hidden;

  &__header {
    @include make-col(10);

    display: flex;
    align-items: flex-end;
    justify-content: space-between;
    margin-bottom: 30px;
  }

  &__box {
    flex-basis: 48.5%;
    max-width: 48.5%;

    .search-box {
      &__label {
        margin-bottom: 17px;
        font-size: 16px;
      }
    }
  }

  &__row {
    @include make-row();

    max-width: 85%;
  }

  &__item {
    @include make-col(6);
    @include make-col-ready();

    margin-bottom: 25px;
  }

  &__top-pagination {
    margin-bottom: 7px;
  }

  &__pagination {
    justify-content: flex-end;
    width: 100%;
  }

  @include media-breakpoint-down(xl) {
    padding-bottom: 62px;

    &__header {
      margin-bottom: 25px;
    }

    &__row {
      max-width: 85.5%;
    }

    &__item {
      margin-bottom: 15px;
    }

    &__pagination {
      margin-top: 13px;
    }
  }

  @include media-breakpoint-down(lg) {
    padding-top: 0;
    padding-bottom: 57px;

    &__header {
      @include make-col(12);

      margin-bottom: 28px;
    }

    &__box {
      flex: 0 0 50.5%;
      max-width: 50.5%;

      .search-box {
        &__label {
          margin-bottom: 9px;
        }
      }
    }

    &__row {
      width: unset;
      max-width: unset;
    }

    &__item {
      margin-bottom: 28px;
    }

    &__pagination {
      justify-content: center;
      margin-top: 0;
    }
  }

  @include media-breakpoint-down(md) {
    padding-bottom: 52px;

    &__box {
      flex: 0 0 48%;
      max-width: 48%;
    }
  }

  @include media-breakpoint-down(sm) {
    padding-bottom: 44px;

    &__header {
      margin-bottom: 13px;
    }

    &__box {
      flex: 0 0 100%;
      max-width: 100%;
    }

    &__top-pagination {
      display: none;
    }

    &__row {
      margin-right: -30px;
      margin-left: -30px;
    }

    &__item {
      @include make-col(12);

      margin-bottom: 17px;
    }

    &__pagination {
      margin-top: 20px;
    }
  }
}
</style>
