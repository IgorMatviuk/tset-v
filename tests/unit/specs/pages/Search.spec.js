import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import VueRouter from 'vue-router'

import {
  mutationTypes as rootMutationTypes,
  mutations as rootMutations
} from '@/store'

import {
  mutations as searchMutations,
  actions as searchActions,
  actionTypes as searchActionsTypes
} from '@/store/search'

import Search from '@/pages/-search'
import BlockSearchResult from '@/components/blocks/BlockSearchResult'
import BlockSearchResultNotFound from '@/components/blocks/BlockSearchResult/BlockSearchResultNotFound'

import productAttrs from '@/api/factories/product'
import createFactory from '@/api/helpers/factory'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(VueRouter)

describe('Search', () => {
  let app
  let store
  let products
  let apiResponse
  let route
  let router
  let mockGetProducts
  let mocksSetBreadcrumbs

  beforeEach(() => {
    products = createFactory(productAttrs, 10)

    apiResponse = {
      data: [...products],
      links: {},
      meta: {
        current_page: 1,
        total: 6
      }
    }

    app = {
      $axios: {
        $get: jest.fn().mockImplementation(() => Promise.resolve(apiResponse))
      },
      i18n: {
        t: (url) => url,
        path: (url) => url
      }
    }

    route = {
      query: {
        q: 'Платья'
      }
    }

    mockGetProducts = searchActions[searchActionsTypes.GET_PRODUCTS].bind(app)
    mocksSetBreadcrumbs = rootMutations[rootMutationTypes.SET_BREADCRUMB]

    store = new Vuex.Store({
      modules: {
        search: {
          namespaced: true,
          state: {
            products: [],
            query: '',
            isLoading: false,
            currentPage: null,
            totalPages: null
          },
          actions: {
            [searchActionsTypes.GET_PRODUCTS]: mockGetProducts
          },
          mutations: { ...searchMutations }
        }
      },
      state: {
        breadcrumb: []
      },
      mutations: {
        [rootMutationTypes.SET_BREADCRUMB]: mocksSetBreadcrumbs
      }
    })

    router = new VueRouter()
  })

  test('has expected structure', async () => {
    const wrapper = shallowMount(Search, {
      localVue,
      store,
      mocks: {
        $scrollTo: () => ''
      }
    })

    await wrapper.vm.$options.fetch({ app, store, route })

    expect(wrapper.element).toMatchSnapshot()
  })

  test('drawing search results', async () => {
    const wrapper = shallowMount(Search, {
      localVue,
      store,
      mocks: {
        $scrollTo: () => ''
      }
    })

    await wrapper.vm.$options.fetch({ app, store, route })
    expect(wrapper.findComponent(BlockSearchResult).exists()).toBe(true)
    expect(wrapper.vm.products.length).toEqual(apiResponse.data.length)
  })

  test('drawing nothing found', () => {
    const wrapper = shallowMount(Search, {
      localVue,
      store,
      mocks: {
        $scrollTo: () => ''
      }
    })

    expect(wrapper.findComponent(BlockSearchResultNotFound).exists()).toBe(true)
    expect(wrapper.vm.products.length).toEqual(0)
  })

  test('change route query', async () => {
    const wrapper = shallowMount(Search, {
      localVue,
      store,
      router,
      mocks: {
        $scrollTo: () => ''
      }
    })

    const query = {
      q: 'брюки'
    }

    wrapper.vm.$router.push({ path: wrapper.vm.$route.path, query })
    await wrapper.vm.$nextTick()
    expect(wrapper.vm.products.length).toEqual(apiResponse.data.length)
  })
})
