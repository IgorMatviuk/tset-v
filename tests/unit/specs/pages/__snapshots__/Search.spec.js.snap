// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search has expected structure 1`] = `
<main
  class="search-page"
>
  <thebreadcrumb-stub
    items=""
  />
   
  <div
    class="search-page__search-result"
  >
    <!---->
     
    <transition-stub
      mode="out-in"
      name="fade"
    >
      <blocksearchresult-stub
        products="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
        query="Платья"
      />
    </transition-stub>
  </div>
</main>
`;
